runtime/src/main/java/org/corfudb/recovery/FastSmrMapsLoader.java

```diff
@@ -183,12 +183,14 @@ public void updateStreamTails(long address, ILogData logData) {
         // On checkpoint, we also need to track the stream tail of the checkpoint
         if (isCheckPointEntry(logData)) {
             if (logData.getCheckpointType() == CheckpointEntry.CheckpointEntryType.END) {
-                streamTails.put(logData.getCheckpointedStreamId(),
-                        getStartAddressOfCheckPoint(logData));
+                streamTails.compute(logData.getCheckpointedStreamId(),
+                        (uuid, value) -> (value == null) ? getStartAddressOfCheckPoint(logData)
```

Comment:
The fastSMRLoader already enforces that in the order of the ranges it processes on bulk reads.